### PR TITLE
fix(runtime): harden trust and reliability boundaries

### PR DESCRIPTION
**Pull request by bymoye**
## Summary

- make migration/schema readiness fail closed and preserve immutable migration history
- replace process-local CSRF state with bounded, session-bound stateless tokens
- add bounded auth snapshots with cancellation-safe single-flight loading, generation-fenced invalidation, and explicit cross-instance invalidation
- enforce one HTTP/WebSocket proxy header policy and explicit trusted-upstream allowlists
- make federation inbox handling durable and idempotent with a lightweight transactional receipt table scoped to each local inbox
- isolate permanently rejected handler writes with a receipt savepoint and fail closed on missing delivery IDs/inboxes
- upgrade the short-lived scope-less receipt table through immutable follow-up migration `013`
- expose only canonical public ActivityPub projections and remove fake collection pagination links
- bound AI task event buffering, coalesce progress/deltas, and preserve terminal control events
- align the backend, proxy, updater and frontend CI gates on locked dependencies and Rust 1.94
- document the fail-closed schema boundary, Move/FileChunk recovery contracts, and complete trusted-proxy deployment examples
- add `AUDIT_REPAIR_TASKS.md` with the evidence-backed disposition of every repair item

The first-owner/setup-chain rebuild is intentionally excluded as requested.

Per #265, this PR preserves the preview-stage Docker Hub tag fallback when GitHub `release.json` is unavailable. Updater source behavior is unchanged; invalid signatures/manifests remain fail closed.

## Validation

- `cargo check --workspace --all-targets --locked`
- `cargo test --workspace --locked` — 1631 passed
- `cargo clippy --workspace --all-targets --locked -- -D warnings`
- focused negative coverage includes:
  - cancelled auth load owners release and wake the single-flight slot
  - invalidation fences stale miss results from re-entering the auth cache
  - opaque signer fallback retains the original signed identity
  - non-positive Accept activity IDs fail closed
  - RoomJoin E2E fanout requires a resolved remote inbox
  - one public activity executes independently for distinct local inbox scopes
- proxy `cargo fmt --all -- --check`
- proxy check/test/Clippy — 34 tests passed
- frontend `pnpm test` — 611 tests plus ESLint, Stylelint and typecheck passed
- frontend `pnpm build`
- PR CI — 7/7 passed

## PostgreSQL integration evidence

- greenfield migrations leave no `schema_check` drift
- a database that recorded old `012` with the scope-less receipt table upgrades through `013` to the authoritative shape
- rolling back to the receipt handler savepoint removes simulated handler writes before a permanent rejection is committed

## Platform boundary

- Updater is Unix-specific; its check/test/Clippy evidence comes from the Ubuntu CI job. Windows compilation is not used as a substitute.

### CODE REVIEW

**Comment by mirai-mamori** [comment]
## 审查反馈（部署影响与取舍）

整体方向认可：CSRF 无状态化、auth 有界缓存、proxy 头策略统一、inbox 幂等、CI `--locked` 都对症。下面是希望合入前对齐的产品/运维取舍。

### 1. 尽量不新增迁移；receipt 优先挂既有主表

- 希望 **尽可能不新增 migration**（本 PR 的 `012_federation_inbox_receipts` 是主要顾虑）。
- Inbox 幂等若必须落库，优先考虑挂在既有联邦主表（例如 `federation_activities` 或紧密相关表）上扩展列/状态，而不是单独建 `federation_inbox_receipts`。
- 接受现实：跨进程/跨重启的权威幂等很难「零 DDL」；若短期做不到主表扩展，宁可保留进程内 dedup + 明确局限，也不要为了完美收据强上新表——除非收益明确且迁移路径极轻。

### 2. Schema / 启动门闸不要过分激进

- 「迁移或 schema repair 失败 → 禁止进入完整服务」对存量/半残库过激，升级风险偏高。
- 更希望分层：
  - **关键路径**（认证、owner、session epoch 等）fail-closed；
  - **非关键漂移** 告警 + 降级，而不是整站拒起；
  - 若保留硬门闸，需要显式、可运维的 bypass，且默认不应一刀切打死流量。

### 3. 联邦能力收窄可以理解，但要写清原因与恢复条件

当前 `Move` / `myriad:FileChunk` → 503 的原因（认可为临时正确）：

| 能力 | 原因 |
| --- | --- |
| **Move** | 校验依赖远程 HTTP 拉 actor；不能在已 claim 的 receipt 事务里夹杂不可回滚的 IO，否则崩溃会留下半截迁移。 |
| **FileChunk** | 依赖文件系统写入，无法与 DB receipt 原子提交；先 503 避免「对端以为成功、本地不一致」。 |

请在 PR/文档里写清：这是收据事务边界下的**临时降级**，以及恢复条件（preflight + 事务内 DB 效果；FileChunk 的事务化 outbox/存储方案）。

### 4. Proxy 信任边界：空 allowlist 的部署冲击必须可落地解决

- 空 `PROXY_TRUSTED_UPSTREAMS` 从「自动信任私网」改为「不信任任何转发头」后，典型 Docker + 宿主 Nginx 会出现客户端 IP / `X-Forwarded-Proto` 不准。
- **运维解法（与现安全模型一致）**：显式配置外层代理在 proxy 容器视角的源地址，例如 `PROXY_TRUSTED_UPSTREAMS=172.17.0.1/32`（以实际为准；禁止 `0.0.0.0/0`）。
- 请同步修正仍写「留空=自动信任私网」的 `.env.production.example`（及同类示例），并考虑 compose 给常见 Docker 宿主网关一个**窄默认**，降低升级踩坑率。
- Backend 侧 `TRUST_PROXY_HEADERS` / `TRUST_PROXY_PEERS` 与 proxy 段是两段链路，文档里建议给一条完整示例。

### 5. 其它已确认、可接受的行为

- CSRF 改为 `v1.` 无状态令牌：旧 32 位 token 失效；官方前端会重拉，可接受。
- Auth 5s 缓存 + NOTIFY：可接受；降权依赖失效/TTL。
- followers/following 去掉假 `first`：可接受（隐私收紧）。
- Updater Docker Hub 回退按 #265 保留：OK。

### 希望的调整优先级

1. **尽量去掉或改写 012 新表迁移**（并进主表 / 更轻的 DDL / 或分阶段）。
2. **放宽「禁止完整服务」**：关键 fail-closed，非关键勿一刀切。
3. **补齐 proxy 升级路径**：示例 env + 可选窄默认 CIDR。
4. **写清 Move/FileChunk 503 的临时性与恢复条件**。

**Comment by bymoye** [comment]
@mirai-mamori 逐项按代码路径、事务边界和部署失败模式复核后，处理结论如下。

1. **保留独立 receipt 表，不并入 `federation_activities`，也不退回进程内 dedup。** `federation_activities` 是部分已接受活动的内容投影，并非所有 inbox handler 都会产生该记录；它无法表达 handler 前的 claim、永久拒绝以及同 ID 异文冲突。进程内 dedup 则无法覆盖多副本、重启及事务回滚，因此不能作为权威幂等边界。考虑到部署成本，`012` 已缩成无数据回写、无二级索引的单表 DDL，并删除了当前事务模型下不可达的 lease、attempts 和 failed 状态。另修复了 receipt key 缺少本地 inbox scope 的问题，避免同一公开活动逐个投递给不同本地用户时只处理第一个 inbox。

2. **不放宽 schema 门闸，也不提供 bypass。** 当前后端是统一路由和统一 schema 的单体服务，没有把 schema capability 隔离到路由；此时无法可靠定义“非关键漂移”。带着已知缺表、缺列或约束漂移继续启动，只会把确定的启动失败转成随机运行期 SQL 错误、部分写入和安全语义漂移。文档已写明安全实现分层启动的前置条件：显式能力声明、受影响路由稳定返回 503、health 暴露能力状态；具备这些机制前继续 fail-closed。

3. **接受 Move / FileChunk 文档意见。** 已补充临时 503 的事务原因及恢复条件：Move 需要事务前完成远端 actor preflight，事务内仅执行可回滚 DB 效果并重验证前置条件；FileChunk 需要事务化 outbox/持久对象存储协议，使 receipt 提交与文件可见性具有可恢复的一致性。

4. **接受 proxy 示例修正，拒绝写死 compose 网关默认值。** 已修正 `.env.production.example`、`SECURITY_HEADERS.md` 和 external-db 示例，并补充外层代理 → proxy、proxy → backend 两段完整信任链。没有默认写入 `172.17.0.1/32`：实际源地址随 Docker 网络、rootless 模式和端口发布实现变化，错误默认既可能使部署失效，也可能扩大信任边界；这里应由部署者按 proxy 实际观测到的直接上游显式配置。

验证：本地 workspace 1625 项测试、all-target check 和 Clippy 全部通过；本 PR 最新 head 的 CI 7/7 通过，包含 PostgreSQL 实库迁移与 schema drift gate。

**Comment by mirai-mamori** [comment]
## 续审（针对最新 tip：`6d03696c` / `8011411e` / `73c89396`）

相对上一轮反馈，下列项已对齐，感谢：

- `.env.production.example` / `SECURITY_HEADERS.md` 纠正空 `PROXY_TRUSTED_UPSTREAMS` 语义，并补齐外层代理 → proxy → backend 两段示例
- receipt 增加 `inbox_scope`，避免同一公开活动只在第一个本地 inbox 执行
- 去掉 lease / `InFlight` / `failed`；未提交 `processing` + PG 唯一键串行，模型更干净
- Move / FileChunk 临时 `503` 的原因与恢复条件已写入文档
- 独立 receipt 表、schema 硬门闸、不写死 compose 网关默认值——接受为有意取舍

下面两条仍建议在合入前修：

### [P2] 修复 `receipt_key` 空 signer 回退仍传错参数

`backend/src/federation/inbox/receipt.rs`（约 66–71 行）：

```rust
let signer = normalize_actor_url(signer);
ReceiptKey {
    signer: if signer.is_empty() {
        signer_from_opaque(signer) // 这里已是 normalize 后的空串
    } else {
        signer
    },
```

`normalize` 清空后，`signer_from_opaque` 只能再得到空串，回退永远无效。应在 normalize 前保留原始输入再 fallback。

### [P2] 让 `012` 能升级旧表形，而不是 `IF NOT EXISTS` 空成功

`backend/migrations/012_federation_inbox_receipts.rs` 在本 PR 内从「`id` + lease + `UNIQUE(signer, activity_id)`」改成了「`PRIMARY KEY (signer, activity_id, inbox_scope)`」，但 `up` 仍只有 `CREATE TABLE IF NOT EXISTS`。

若某环境已跑过中间版 tip：迁移名已记录、表形仍是旧的 → 缺 `inbox_scope` → 运行期 SQL 失败，再被 schema fail-closed 挡住。合入 `preview` 的全新库没事；修过本分支的持久化 DB 会踩坑。

建议 `up` 检测旧形并 `DROP`/重建或 `ALTER` 对齐，不能只靠 `IF NOT EXISTS`。

---

其余 CSRF / auth cache / proxy 信任策略可按你回复中的结论保留。修完上述两点后，从代码审查侧可以再看一眼合入。

File: backend/src/middleware/auth.rs [hunk 1]
```diff
@@ -273,30 +453,96 @@ pub async fn verify_current_admin_from_headers(
...
-    }))
+
+    ensure_auth_cache_listener(db).await;
+
+    loop {
+        if let Some(snapshot) = auth_cache_get(user_id) {
+            return Ok(snapshot);
+        }
+
+        // Only one request per user performs the miss query. Other concurrent
```
Comment (mirai-mamori) [inline backend/src/middleware/auth.rs:551]: **[bug]** Auth-cache single-flight still leaks the in-flight map entry when the owner task is cancelled (or panics) between `in_flight.insert(user_id, …)` and the later `remove`. Waiters then block forever on `receiver.changed().await` for that `user_id` until process restart. There is no `Drop`/scope guard that clears and notifies waiters on cancellation.

**Suggestion:** Hold an RAII guard (or `scopeguard`) created at insert time that always removes the entry and `send`s on drop; only the successful owner path should also `auth_cache_put`. Ensure waiter cancellation does not leave orphans either (owner owns cleanup exclusively).

File: backend/src/middleware/auth.rs [hunk 2]
```diff
@@ -273,30 +453,96 @@ pub async fn verify_current_admin_from_headers(
...
+                [user_id.into()],
+            ))
+            .await
+            .map(|row| {
+                row.map(|row| AuthSnapshot {
+                    token_version: row
+                        .try_get::<i32>("", "token_version")
+                        .ok()
+                        .map(i64::from)
+                        .or_else(|| row.try_get::<i64>("", "token_version").ok())
```
Comment (mirai-mamori) [inline backend/src/middleware/auth.rs:580]: **[bug]** `auth_cache_put` still races with `invalidate_auth_cache_local` / NOTIFY invalidation. A miss-path load that started before demotion, `token_version` bump, or account deletion can finish after invalidate and re-insert a stale positive snapshot for a full new `AUTH_CACHE_TTL` (5s). Local invalidate therefore does not establish a hard upper bound under concurrency.

**Suggestion:** Per-user generation/tombstone: capture generation at miss start; bump generation on every local invalidate; `put` only if generation is unchanged. Optionally refuse to put over a tombstone until the next successful live load after invalidate.

File: backend/src/federation/inbox/receive.rs [hunk 3]
```diff
@@ -1834,6 +2110,49 @@ async fn enqueue_delivery(
...
+                activity.id.clone().into(),
+                user_id.into(),
+                activity.activity_type.clone().into(),
+                activity_json.into(),
+            ],
+        ))
+        .await
+        .map_err(db_err)?;
+    let act_id: i32 = act_row
+        .map(|r| r.try_get("", "id").unwrap_or(0))
```
Comment (mirai-mamori) [inline backend/src/federation/inbox/receive.rs]: **[bug]** `enqueue_delivery_queue` still fail-opens when the activity insert does not yield an id: `try_get(...).unwrap_or(0)` / `unwrap_or(0)`. A missing/unreadable `RETURNING id` produces `activity_id = 0` in `federation_delivery_queue`, so the Accept may never be deliverable while the inbound Follow receipt is still committed as accepted (peer stops retrying).

**Suggestion:** Fail closed: require a positive `act_id` (`ok_or` / map to `db_err` or 500) and return `Err` so the receipt transaction rolls back.

File: backend/src/federation/room/inbox.rs [hunk 4]
```diff
@@ -822,8 +815,9 @@ pub(crate) async fn refanout_local_e2e_keys_to_member(
         .await
         .map_err(|e| e.to_string())?;
     let Some(target) = target else {
-        // Try fetch once
-        let _ = crate::federation::actor::fetch_remote_actor(db, target_actor).await;
+        // Do not perform remote HTTP actor fetches while an inbound receipt
+        // transaction is open.  The normal actor/discovery path can populate
```
Comment (mirai-mamori) [inline backend/src/federation/room/inbox.rs]: **[bug]** `refanout_local_e2e_keys_to_member` still returns `Ok(())` when the target has no `remote_actors` row. Remote HTTP fetch was removed (correct for open receipt transactions), but the join path also stopped ensuring the actor exists. With durable receipts, a successful RoomJoin that cannot resolve the new member’s inbox is accepted permanently; E2E key fanout is skipped and peers will not retry.

**Suggestion:** Preflight `fetch_remote_actor` for the joining member before claiming the receipt (same pattern as Follow), or return a retryable 503 when the inbox cannot be resolved inside the transaction so the peer redelivers after discovery. Do not commit an accepted receipt that silently drops required KeyExchange fanout.

File: backend/src/federation/inbox/receive.rs [hunk 5]
```diff
@@ -103,24 +181,145 @@ pub async fn post_inbox(
...
+    .await;
+    match result {
+        Ok(status) => finish_and_commit(txn, &key, ReceiptOutcome::Accepted, status, None).await,
+        Err((status, body)) if receipt_result_is_permanent(status) => {
+            let message = body
+                .0
+                .get("error")
+                .and_then(|v| v.as_str())
+                .unwrap_or("permanent federation inbox rejection")
+                .to_string();
```
Comment (mirai-mamori) [inline backend/src/federation/inbox/receive.rs:335]: **[suggestion]** On permanent 4xx handler results, `finish_and_commit(..., ReceiptOutcome::Rejected, …)` commits the entire open transaction, including any handler DB writes performed before the error. The durable receipt then records `rejected` while partial side effects remain, and the peer will not retry.

**Suggestion:** For Rejected outcomes, either (a) roll back handler writes and commit only the rejected receipt in a fresh/savepoint-isolated transaction, or (b) guarantee handlers perform all permanent validation before any write and add tests that assert zero side effects on rejected receipts.

**Review by mirai-mamori — CHANGES_REQUESTED** [review]
## Summary
This PR hardens several runtime trust boundaries (fail-closed schema startup, proxy hop-by-hop/forwarded-header policy, stateless CSRF, auth snapshot cache, durable federation inbox receipts, AI task mailbox bounds, and public ActivityPub object projection). The durable receipt redesign correctly removes the prior in-process InFlight→202 data-loss path and documents temporary 503s for Move/FileChunk. Residual correctness risks remain in the auth single-flight/cache put path (cancel leak + invalidate TOCTOU), fail-open `act_id = 0` Accept queueing, room E2E key fanout that still returns success when the remote inbox is unknown (now worse under durable accept), and permanent-4xx receipt completion that can commit partial handler writes.

## Issue counts by severity

- bugs: 4
- suggestions: 1
- nits: 0

## Prior-issue disposition

1. **Auth single-flight cancel leak** — **still open**. In-flight entry is still removed only on the success path after the DB await (`auth.rs` insert ~476 / remove ~519); no cancel-safe guard.
2. **Auth cache put TOCTOU after invalidate** — **still open**. Unconditional `auth_cache_put` after load (~516) with no generation/tombstone.
3. **Federation InFlight → HTTP 202** — **fixed**. Process-local InFlight map is gone; `claim_receipt` uses transactional INSERT + `FOR UPDATE`, and concurrent losers wait for a committed `accepted`/`rejected` (or see a rolled-back claim and re-execute). `AlreadyAccepted` only after commit.
4. **Room E2E key fanout silent `Ok(())`** — **still open** (worsened under durable accept). Still returns `Ok(())` when remote actor/inbox is missing (~819) and no longer fetches/populates inside the handler.
5. **`act_id = 0` fail-open in `enqueue_delivery_queue`** — **still open**. `unwrap_or(0)` remains (~2141).
6. **Permanent 4xx `finish_and_commit` partial writes** — **still open**. Permanent branch still commits the same txn as handler effects (~257).

**Comment by bymoye** [comment]
@mirai-mamori 这两点均已处理。signer 回退确实传错了值，现保留 normalize 前原始输入，并加了 opaque signer 负向测试。旧 012 表形风险也成立，但直接改 012 的 up 不能修复已写入 seaql_migrations 的数据库，因为迁移器不会重跑同名迁移；因此保留 012 历史，新增 013_federation_inbox_receipts_v2 专门识别并重建短期 scope-less 表形。旧表无法可靠推导 inbox_scope，伪造 scope 会再次造成跨 inbox 误抑制，所以不做虚假的数据映射。最新 PostgreSQL CI 已实际模拟旧 012 表、撤销仅 013 的账本记录并完成升级，最终 schema drift 为零。正式 review 的五项问题也已同步修复，见各 inline 回复；当前 CI 7/7 通过。

File: backend/src/middleware/auth.rs (hunk 1 above)
Comment (bymoye) [reply backend/src/middleware/auth.rs:551]: 已修复。in-flight 所有权改为跨 DB await 持有的 RAII guard；Drop 使用 sender 的 Arc 身份确认后清理槽位并通知 waiter，因此 owner task 被 abort 或 panic 时也不会留下永久占位。新增负向测试直接模拟 owner drop，验证 waiter 立即结束等待且同一 user_id 可重新取得 owner。

File: backend/src/middleware/auth.rs (hunk 2 above)
Comment (bymoye) [reply backend/src/middleware/auth.rs:580]: 已修复。采用 AuthCache 内的有界全局 monotonic generation，而非随 user_id 无界增长的 per-user map：miss 开始时捕获 generation，invalidate 在同一互斥锁内递增 generation 并删除 entry，put 也在该锁内比较后写入。这样 invalidate 前开始的查询无法在 invalidate 后重新插入；无关失效最多保守跳过一次缓存填充。对应竞态负向测试已加入。

File: backend/src/federation/inbox/receive.rs (hunk 3 above)
Comment (bymoye) [reply backend/src/federation/inbox/receive.rs]: 已修复。INSERT RETURNING 现在必须返回一行、id 必须可读取且严格大于 0；任一条件不满足都会返回 Err，使 Accept outbox 与 inbound receipt 的同一事务整体回滚。新增 0 和负数 ID 的 fail-closed 测试。

File: backend/src/federation/room/inbox.rs (hunk 4 above)
Comment (bymoye) [reply backend/src/federation/room/inbox.rs]: 已修复。RoomJoin 在 claim 前解析真正 joining member 并完成远端 actor preflight；事务内不再发起 HTTP。缺失或空 inbox 现在返回错误，KeyExchange activity/outbox 必须成功入队，fanout 错误由 handle_room_join 向上传播，并且在任何 realtime notification 之前执行，因此 membership 与 fanout 会一起回滚并让对端重试。新增缺失/空 inbox 负向测试。

File: backend/src/federation/inbox/receive.rs (hunk 5 above)
Comment (bymoye) [reply backend/src/federation/inbox/receive.rs:335]: 已修复，采用建议中的 savepoint 隔离。receipt claim 后、handler 前建立 savepoint；永久 4xx 先 ROLLBACK TO SAVEPOINT 清除全部 handler DB 写入，再在仍持有 claim/唯一键的同一事务中写 rejected receipt 并提交，避免 fresh transaction 的并发抢占窗口。PostgreSQL CI 用实际事务写入探针验证回滚后副作用行数为 0。
